File: node/src/domains/idempotency.js
// idempotency — replay-safe writes per the IETF Idempotency-Key shape (the Stripe pattern). The dangerous
// property is EXACTLY-ONCE: same key + same body returns the STORED response (the side effect never re-runs);
// same key + a different body is a 409; no key means no deduplication (opt-in, per the standard). The claim is
// ONE atomic read-modify-write through storeDo — two processes racing the same key produce exactly one winner;
// the loser is served the winner's stored response. Durable: a replay works after a restart.
import { isStrictInt, nextId, problem, requireIdentity, sendJSON, storeGet } from '../core/runtime.js';
import { digestHex, scopedKey } from '../parts/digest.js';
import { claimOnce } from '../parts/idempotent_claim.js';
import { isWellFormed } from '../parts/well_formed.js';

// state in store: seq "idempotency_payment" · ns "idempotency_keys" SCOPED-key -> {id, amount, body_hash, caller}.
// The store key is SCOPED TO THE CALLER (scopedKey below) so an Idempotency-Key is PRIVATE to its caller.

const ROUTE = 'POST /idempotency/payments'; // the route discriminator (per-route slot, GAP-6)

// body_hash = the FULL request body fingerprint (here the one field amount) — the SAME-KEY-DIFFERENT-BODY guard,
// SEPARATE from the lookup key. A copier whose body gains fields MUST add them here.
const digest = (amount) => digestHex('amount', amount); // the central canonical fingerprint (digest part)

export async function idempotencyPay(req, res, params, body) {
  // identity: the caller must be AUTHENTICATED (deny-by-default, no token -> 401) BEFORE validation and before the
  // Idempotency-Key. The Idempotency-Key is a DEDUPE token, NOT identity — kept ON TOP of authn AND the dedup slot is
  // SCOPED TO THE CALLER (the key is private to its caller).
  const caller = await requireIdentity(req, res);
  if (caller === null) return;
  if (!body || !isStrictInt(body, 'amount') || body.amount < 1) {
    return problem(res, 422, 'amount must be a positive integer');
  }
  const key = req.headers['idempotency-key'];
  if (key === undefined) {
    // no key -> no dedupe; every request is a fresh side effect
    return sendJSON(res, 201, { id: await nextId('idempotency_payment'), amount: body.amount });
  }
  // an Idempotency-Key is a SINGLE opaque token; node comma-joins duplicate headers into one string, so count the
  // raw header lines and REJECT >1 (deterministic + IDENTICAL ×3 — go/python also reject), never dedupe on the
  // ambiguous joined value.
  let nKeys = 0;
  for (let i = 0; i < req.rawHeaders.length; i += 2) {
    if (req.rawHeaders[i].toLowerCase() === 'idempotency-key') nKeys += 1;
  }
  if (nKeys > 1) {
    return problem(res, 422, 'Idempotency-Key must be a single value');
  }
  if (!isWellFormed(key)) {
    return problem(res, 422, 'Idempotency-Key must be non-empty with no control characters');
  }
  const h = digest(body.amount);
  const scoped = scopedKey(ROUTE, caller, key); // the central caller-scoped, collision-safe slot (digest part)
  let prior = await storeGet('idempotency_keys', scoped); // fast path: a settled key never mints
  if (prior === undefined) {
    // mint BEFORE the claim (a race loser's id is a gap), then claim atomically via the central part
    const rec = { id: await nextId('idempotency_payment'), amount: body.amount, body_hash: h, caller };
    prior = await claimOnce('idempotency_keys', scoped, rec);
  }
  if (prior.caller !== caller) {
    // DEFENSE-IN-DEPTH: the scoped key already isolates callers; a stored-caller mismatch is structurally impossible,
    // so if it ever happens (collision / regression) REFUSE rather than cross-replay.
    return problem(res, 409, 'idempotency key is not owned by this caller');
  }
  if (prior.body_hash !== h) {
    return problem(res, 409, 'idempotency key reused with a different body');
  }
  // first call and every replay: the SAME response
  sendJSON(res, 201, { id: prior.id, amount: prior.amount });
}

File: go/internal/domains/idempotency/idempotency.go
// Package idempotency — replay-safe writes per the IETF Idempotency-Key shape (the Stripe pattern). The
// dangerous property is EXACTLY-ONCE: same key + same body returns the STORED response (the side effect never
// re-runs); same key + a different body is a 409; no key means no deduplication (opt-in, per the standard).
// The claim is ONE atomic read-modify-write through (*KV).Do — two processes racing the same key produce
// exactly one winner; the loser is served the winner's stored response. Durable: a replay works after restart.
package idempotency

import (
	"net/http"

	"app/internal/core"
	"app/internal/parts/digest"
	"app/internal/parts/idempotent_claim"
	"app/internal/parts/well_formed"
)

type idempotencyRec struct {
	Id       int    `json:"id"`
	Amount   int    `json:"amount"`
	BodyHash string `json:"body_hash"`
	Caller   string `json:"caller"` // the OWNING caller — the dedup slot is private to it (defense-in-depth)
}

var idempotencyKeys = core.NewKV[string, idempotencyRec]("idempotency_keys")

const idempotencyRoute = "POST /idempotency/payments" // the route discriminator (per-route slot, GAP-6)

func idempotencyDigest(amount int) string {
	// body_hash = the FULL request body fingerprint (here the one field amount) — the SAME-KEY-DIFFERENT-BODY guard,
	// SEPARATE from the lookup key. A copier whose body gains fields MUST add them here.
	return digest.DigestHex("amount", amount) // the central canonical fingerprint (digest part)
}

func IdempotencyPay(w http.ResponseWriter, r *http.Request) {
	// identity: the caller must be AUTHENTICATED (deny-by-default, no token -> 401) BEFORE decode and before the
	// Idempotency-Key. The Idempotency-Key is a DEDUPE token, NOT identity — kept ON TOP of authn AND the dedup slot
	// is SCOPED TO THE CALLER (the key is private to its caller).
	caller, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	// decode FIRST: DecodeJSON enforces the body cap (413) and drains the stream before any reply
	in, ok := core.DecodeJSON[struct {
		Amount *int `json:"amount"`
	}](w, r)
	if !ok {
		return
	}
	if in.Amount == nil || *in.Amount < 1 || *in.Amount > core.MaxSafeInt { // bounded to the ×3-safe range
		core.WriteProblem(w, 422, "amount must be a positive integer")
		return
	}
	key, hasKey := r.Header["Idempotency-Key"]
	if !hasKey { // no key -> no dedupe; every request is a fresh side effect
		pid := core.NextID("idempotency_payment")
		core.WriteJSON(w, 201, map[string]any{"id": pid, "amount": *in.Amount})
		return
	}
	if len(key) > 1 { // an Idempotency-Key is a SINGLE opaque token; duplicate headers are ambiguous -> reject (×3 parity)
		core.WriteProblem(w, 422, "Idempotency-Key must be a single value")
		return
	}
	k := key[0] // exactly one value (hasKey -> len>=1; duplicates rejected above)
	if !well_formed.IsWellFormed(k) { // a PRESENT key must be a well-formed identifier
		core.WriteProblem(w, 422, "Idempotency-Key must be non-empty with no control characters")
		return
	}
	digestVal := idempotencyDigest(*in.Amount)
	scoped := digest.ScopedKey(idempotencyRoute, caller, k) // the central caller-scoped, collision-safe slot (digest part)
	prior, settled := idempotencyKeys.Get(scoped) // fast path: a settled key never mints
	if !settled {
		// mint BEFORE the claim (a race loser's id is a gap), then claim atomically via the central part
		rec := idempotencyRec{Id: core.NextID("idempotency_payment"), Amount: *in.Amount, BodyHash: digestVal, Caller: caller}
		prior = idempotent_claim.ClaimOnce(idempotencyKeys, scoped, rec)
	}
	if prior.Caller != caller {
		// DEFENSE-IN-DEPTH: the scoped key already isolates callers; a stored-caller mismatch is structurally
		// impossible, so if it ever happens (collision / regression) REFUSE rather than cross-replay.
		core.WriteProblem(w, 409, "idempotency key is not owned by this caller")
		return
	}
	if prior.BodyHash != digestVal {
		core.WriteProblem(w, 409, "idempotency key reused with a different body")
		return
	}
	// first call and every replay: the SAME response
	core.WriteJSON(w, 201, map[string]any{"id": prior.Id, "amount": prior.Amount})
}
